File: upnext-ios/UpNext/UpNext/UpHeroAudio.swift
import Foundation

/// Audio follows log events, so automatic and manual skill activation share the same cue.
enum UpHeroAudio {
    static func music(_ session: CombatSession?) -> MusicTrack {
        guard let session else { return .main }
        let dungeon = MusicTrack(rawValue: session.dungeonId.rawValue) ?? .main
        if session.status == .completed { return dungeon }
        for entry in session.log.reversed() {
            switch entry {
            case .boss: return .boss
            case let .encounter(monster, _): return monster.isBoss == true ? .boss : dungeon
            case .victory, .floor, .sessionEnd: return dungeon
            default: continue
            }
        }
        return dungeon
    }

    static func cue(_ entry: LogEntry) -> SoundName? {
        switch entry {
        case let .combat(attacker, damage, outcome, _, _, _, _):
            switch outcome {
            case .crit: return .criticalHit
            case .dodge: return .dodge
            case .miss: return .miss
            case .hit: return damage > 0 ? (attacker == .hero ? .heroHit : .enemyHit) : .shieldBlock
            }
        case let .encounter(monster, _): return monster.isBoss == true ? nil : .encounter
        case let .victory(monster, _, _, _, _, _): return monster.isBoss == true ? .fullClear : .battleWin
        case let .drop(equipment, _): return [Rarity.unique, .legend].contains(equipment.rarity) ? .rareLoot : .lootDrop
        case .treasure: return .treasure
        case .floor: return .floorAdvance
        case let .choice(_, _, _, _, resolved, _, _, _, _, _): return resolved == nil ? .choiceOpen : nil
        case let .skill(classType, _, _, _, _, _, _):
            switch classType {
            case .priest, .druid: return .skillHoly
            case .chronomancer: return .skillTime
            case .mage, .illusionist, .bard: return .skillMagic
            default: return .skillMelee
            }
        case let .monsterEffect(effect, _, _, _, _, _):
            switch effect {
            case .regen: return .heal
            case .poisonTick: return .poison
            case .shieldBlock: return .shieldBlock
            }
        case let .choiceResult(_, _, summary, _, _, _, _, slot, _):
            if slot != nil { return nil }
            if (summary?.damage ?? 0) > 0 { return .enemyHit }
            if (summary?.heal ?? 0) > 0 { return .heal }
            if (summary?.coins ?? 0) != 0 || (summary?.xp ?? 0) != 0 { return .rewardChoose }
            return .confirm
        case let .sessionEnd(reason, _, _, _, _, _, _):
            switch reason {
            case .heroDied, .defeat: return .defeat
            case .timeExpired: return .timeWarning
            case .heroAbandoned, .abandoned: return .retreat
            default: return nil
            }
        default: return nil
        }
    }
}

File: src/lib/upHeroAudio.ts
import type { CombatSession, LogEntry } from '@/types/uphero';
import type { MusicTrack } from './audio';
import type { SoundName } from './audioCatalog';

/** Inspect encounter lifetime, not floor number: choices and paused banners retain boss music. */
export function dungeonMusic(session: CombatSession | null): MusicTrack {
  if (!session) return 'main';
  if (session.status === 'completed') return session.dungeonId;
  for (let i = session.log.length - 1; i >= 0; i--) {
    const entry = session.log[i];
    if (entry.type === 'boss') return 'boss';
    if (entry.type === 'encounter') return entry.monster.isBoss ? 'boss' : session.dungeonId;
    if (entry.type === 'victory' || entry.type === 'floor' || entry.type === 'sessionEnd') break;
  }
  return session.dungeonId;
}

export function combatSound(entry: LogEntry): SoundName | null {
  switch (entry.type) {
    case 'combat':
      if (entry.outcome === 'crit') return 'criticalHit';
      if (entry.outcome === 'dodge') return 'dodge';
      if (entry.outcome === 'miss') return 'miss';
      return entry.damage > 0 ? entry.attacker === 'hero' ? 'heroHit' : 'enemyHit' : 'shieldBlock';
    case 'encounter': return entry.monster.isBoss ? null : 'encounter';
    case 'victory': return entry.monster.isBoss ? 'fullClear' : 'battleWin';
    case 'drop': return ['unique', 'legend'].includes(entry.equipment.rarity) ? 'rareLoot' : 'lootDrop';
    case 'treasure': return 'treasure';
    case 'floor': return 'floorAdvance';
    case 'choice': return entry.resolvedIndex == null ? 'choiceOpen' : null;
    case 'skill':
      if (['priest', 'druid'].includes(entry.classType)) return 'skillHoly';
      if (entry.classType === 'chronomancer') return 'skillTime';
      if (['mage', 'illusionist', 'bard'].includes(entry.classType)) return 'skillMagic';
      return 'skillMelee';
    case 'monsterEffect': return entry.effect === 'regen' ? 'heal' : entry.effect === 'poisonTick' ? 'poison' : 'shieldBlock';
    case 'choiceResult':
      if (entry.slot) return null; // The reels own their reveal sound.
      if (entry.effectSummaryData?.damage) return 'enemyHit';
      if (entry.effectSummaryData?.heal) return 'heal';
      if (entry.effectSummaryData?.coins || entry.effectSummaryData?.xp) return 'rewardChoose';
      return 'confirm';
    case 'sessionEnd':
      if (entry.reason === 'heroDied' || entry.reason === 'defeat') return 'defeat';
      if (entry.reason === 'timeExpired') return 'timeWarning';
      if (entry.reason === 'heroAbandoned' || entry.reason === 'abandoned') return 'retreat';
      return null; // Boss victory already owns the fanfare.
    default: return null;
  }
}
